### PR TITLE
Add `decoration` parameter to `Window` and `DialogWindow` to control the thickness of the undecorated window resizers.

### LINKED ISSUE #4574: Make resize area larger or configurable for undecorated windows
When making a window undecorated, the area that you can click in the corners on macOS is sigificantly smaller than when the window is decorated. This leads to frustration as you are tying to get the mouse into the tiny little area that allows you to click and drag to resize an undecorated window.

I would like to make this area larger. I am not sure if this is more on the compose side or the swing side.

Platform: Desktop
OS: macos 14.4.1 arm64
Compose Version: 1.6.1
Kotlin version: 2.0.0-Beta5

* note that an undecorated window also shows different icons when resizing compared to a decorated window, at least on my platform

### PR DESCRIPTION
**Pull request by m-sasha**
The thickness of the undecorated window resizers is currently hardcoded to 8.dp. This PR adds a composition local, `LocalUndecoratedWindowResizerBorderThickness` to allow users to change it.

Fixes https://github.com/JetBrains/compose-multiplatform/issues/4574

## Testing
Tested manually with:
```
fun main() = application {
    var thickBorders by remember { mutableStateOf(false) }
    Window(
        onCloseRequest = ::exitApplication,
        decoration = when {
            thickBorders -> WindowDecoration.Undecorated(40.dp)
            else -> WindowDecoration.Undecorated()
        }
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier
                .fillMaxSize()
                .padding(100.dp)
        ) {
            Checkbox(
                checked = thickBorders,
                onCheckedChange = { thickBorders = it },
            )
            Text("Thick resizers")
        }
    }
}
```

It's difficult to write an automated test for this because `UndecoratedWindowResizer` relies on `MouseInfo.getPointerInfo()` which can't be simulated from tests.

### Features - Desktop
The thickness of border resizers in undecorated windows and dialogs can now be controlled by passing a new `decoration` argument.

### CODE REVIEW

**Review by MatkovIvan — COMMENTED** [review]
[API Guidelines](https://github.com/androidx/androidx/blob/androidx-main/compose/docs/compose-component-api-guidelines.md#explicit-vs-implicit-dependencies):
> Avoid implicit inputs provided via `CompositionLocal`

**Comment by m-sasha** [comment]
> [API Guidelines](https://github.com/androidx/androidx/blob/androidx-main/compose/docs/compose-component-api-guidelines.md#explicit-vs-implicit-dependencies):
> 
> > Avoid implicit inputs provided via `CompositionLocal`

We don't currently expose `UndecoratedWindowResizer`, so there is no place to pass an explicit argument.
We could expose it, or provide an additional parameter to `Window` and `DialogWindow`, but each of those comes with its own considerations and downsides.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/UndecoratedWindowResizer.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/UndecoratedWindowResizer.desktop.kt]: > We could expose it, or provide an additional parameter to Window and DialogWindow, but each of those comes with its own considerations and downsides.

This way is more friendly for users. I would also stick to it, even if it is harder to design. If we add a local for every configuration, the API will quickly become hard to understand.

We can add something like this (this is an example, I haven't think about it thoroughly):
```
@ExperimentalComposeUIApi Window(undecoratedFrame = WindowUndecoratedFrame(resizerBorderThickness = borderThickness)
```
(and create an issue to remove experimental in 1.8)

**Review by igordmn — CHANGES_REQUESTED** [review]

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/UndecoratedWindowResizer.desktop.kt
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/UndecoratedWindowResizer.desktop.kt]: > If we add a local for every configuration, the API will quickly become hard to understand.

I don't think it will be hard to understand - what is hard about a simple composition local providing a value?

The downside of a composition local is that it's harder to discover. The downside of an explicit parameter is API bloat, and looking at `Window`, it already has 14 parameters.

Additionally, changing the thickness of the undecorated resizer is rather niche, so I'm not sure it's good to have it as an explicit argument.

Finally, it doesn't work well with the current `undecorated` argument because it's not orthogonal. You can specify `undecorated = false` (or leave it the default) and `undecoratedFrame = ...` and then it's not clear what happens. We could add another overload to deal with that. Something like:
```
@Composable
Window(
    ...
    decoration: WindowDecoration = WindowDecoration.SystemDecoration
)

interface WindowDecoration {
    object SystemDecoration : WindowDecoration
    class Undecorated(val resizerThickness: Dp = DefaultWindowResizerThickness) : WindowDecoration
}
```

I'm not strongly against doing this, but API bloat is something to consider here...

Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/UndecoratedWindowResizer.desktop.kt]: > what is hard about a simple composition local providing a value?

As you mentioned, it is hard to discover, this is a big disadvantage.

A `Local` has an advantage that it affects all children, but it can be solved by providing a custom wrapper `MyAppWindow`.

> undecorated resizer is rather niche

We can say that about 90% of parameters 🙃

> The downside of an explicit parameter is API bloat,

If we don't want to bloat API of a single function, we can decompose it to low-level functions. If user needs customization, they go to a lower level:
```
Window(undecorated = true, resizable = false) {
  UndecoratedWindowResizer {

  }
}
```
On the other hand, if we think that `UndecoratedWindowResizer` is inseparable part of `Window`, then the API bloat is inevitable. A `Local` just hides this fact, doesn't eliminate it.

> Finally, it doesn't work well with the current undecorated argument because it's not orthogonal.

The same is for Local - it doesn't work if `undecorated = false`. With parameters on the other hand, we can't provide different overloads as you mentioned.

Overall, I don't think that `undecorateFrame` non-working in `Window(undecarated = false, undecoratedFrame = ` call is an issue, as we explicitly say that `Frame` is `undecorated`.

Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/UndecoratedWindowResizer.desktop.kt]: Your suggestion with a new overload `WindowDecoration` looks good.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt]: Better to use `DeprecationLevel.WARNING` and provide `ReplaceWith`

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]:
```suggestion
sealed class WindowDecoration {
```

AOSP uses `sealed class` in a case when it isn't possible to have custom implementation.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: Maybe we delete it? Users will use `WindowDecoration.System` or `WindowDecoration.Undecorated` directly. I can't think of a case when they need to pass a bool.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: If we decide to add customisation in the future (show/hide some buttons, hide titlebar), can we convert it to `class System`?

I think we can preserve binary compatibility, but source compatibility will be broken (we can live with some source incompatibility though).

Alternative names to consider: `Default`, `SystemDefault`. Choose what you think is good.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: I thought we didn't want to use `sealed` or `enum` classes because it can make it hard to provide backwards compatibility later when we want to add additional subtypes. For example, someone can write
```
when (decoration) {
    is System -> ...
    is Undecorated -> ...
}
```
and when we add another type, this will be broken.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt]: The reason I want HIDDEN is that with WARNING, code that specifies neither `undecorated` nor `decoration` becomes ambiguous:

image

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: It's a function I needed anyway, and I thought I should make it public to make the transition for users simpler.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: Renamed to `SystemDefault`. If we want to add customizations we can add a class `System`.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: Why do you need it?

I wouldn't add a new public API just for migration, users can use another code for migration:
```
if (undecorated) WindowDecoration.Undecorated() else WindowDecoration.System
```

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: Right 🤔, let's keep `interface`

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: I need it just to avoid repeating that line in several places. We don't have to make a public API out of it, of course.

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: Changed it to be `internal`.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt]: We can make `undecorated` required in the Deprecated overload, it will probably help.

**Review by igordmn — APPROVED** [review]
Looks good 👍. Let's wait for the second approval.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: Agree with @igordmn - it should be `sealed`, but not `class`, `sealed interface` should work here. I guess it's covered by some policy - need to look for formal rule there, but the idea is to avoid implementation outside of our code. Explicit requirement in case of adding is pro, not cons

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: I'd move it from global scope to companion object

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: > Agree with @igordmn - it should be sealed, but not class, sealed interface should work here. I guess it's covered by some policy - need to look for formal rule there, but the idea is to avoid implementation outside of our code. Explicit requirement in case of adding is pro, not cons

If we want a `sealed` interface here, we need to hide the actual subtypes from the user, to make the `when` scenario I mentioned impossible.

Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: > I'd move it from global scope to companion object

We can reuse the AOSP pattern:

```
object WindowDecorationDefaults {
    val ResizerThickness: Dp = 8.dp
}
```

It is more preferable over companion object, because it will be consistent with AOSP

Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: On the other hand, it is only for `Undecorated`.

I delegate the decision to @m-sasha

**Review by igordmn — APPROVED** [review]

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: > we need to hide the actual subtypes from the user, to make the when scenario I mentioned impossible.

Why do we need to hide it? Why do we want to make such switch/when impossible?

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: Because adding a new subtype will break it.

Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: > Explicit requirement in case of adding is pro, not cons

It's OK to require extra case handling in this case. It won't break _binary_ compatibility that we **have to** maintain

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (Sanlorng) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: > Agree with @igordmn - it should be `sealed`, but not `class`, `sealed interface` should work here. I guess it's covered by some policy - need to look for formal rule there, but the idea is to avoid implementation outside of our code. Explicit requirement in case of adding is pro, not cons

If we design this interface similar to `TextFieldDecoration `and move the implementation of the `Undecorated Resizer` to this part, it can be provided to users to achieve an immersive title bar application experience on Windows. This requires exposing the `Window `to set window properties and exposing the `HitTest `related interfaces for hit testing of controls within the title bar. You can refer to this [existing implementation](https://github.com/Konyaco/compose-fluent-ui/pull/57/commits/b5f6aa3888d6b6f6295e0569ea0355bd9c09d50b) for specifics. Additionally, in the current Resizer implementation, if the user calls a `Dialog `or `Popup`, the Resizer response will be blocked and won’t respond . If we implement the Resizer through this new interface and always place the `WindowDecoration `field above the `ContentLayer`, can this solve the problem? Would this design make the Decoration API more user-friendly and extensible?

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: I'd rather not dive into the exact semantics of what **binary** compatibility means.

De-facto, valid code that was written and compiled before the addition of another subtype will crash at runtime afterwards. That's breaking existing code.

Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: Yes, sure. But the policy for breaking source code compatibility is MUCH softer than for binary compatibility. The reason behind this is affecting 3rd party libraries that users cannot control.
It doesn't sound like a reason for avoid exposing subtypes at all

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: It's not source compatibility that will be broken. Existing compiled code will throw an exception at runtime.

Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt]: All of it is about possibilities and potential changes. Not a blocker at the moment I guess, we can discuss it when we'll really need this

**Review by MatkovIvan — APPROVED** [review]

**Pull request merged**
